pkg/storegateway/bucket.go

```diff
@@ -1560,39 +1495,39 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 		}
 	}
 
-	s.blocksMx.RLock()
-
-	var mtx sync.Mutex
+	var setsMtx sync.Mutex
 	var sets [][]string
-	for _, b := range s.blocks {
-		b := b
-
+	s.blockSet.all(func(b *bucketBlock) {
 		if !b.overlapsClosedInterval(req.Start, req.End) {
-			continue
+			return
 		}
 		if len(reqBlockMatchers) > 0 && !b.matchLabels(reqBlockMatchers) {
-			continue
+			return
 		}
 
 		resHints.AddQueriedBlock(b.meta.ULID)
 
+		// This index reader shouldn't be used for ExpandedPostings, since it doesn't have the correct strategy.
+		// It's here only to make sure the block's indexReader is held open inside the goroutine below.
+		indexr := b.loadedIndexReader(ctx, selectAllStrategy{}, stats)
```

Comment:
this means we lazy-load the blocks sequentially instead of concurrently. Can we find a way to avoid this? Perhaps a variation of `all` (or `forEach`), let's call it, `borrowBlock`. `borrowBlock` increments `b.pendingReaders` (directly or indirectly) and passes a `func()` to the callback. Calling the `func()` will invoke `b.pendingReaders.Done()` (or .Done **is** the `func()`).

Reply:
It feels a bit too complicated for what we are doing here. How about we segregate the creation of the `indexReader` and the warm-up of the blocks index header — there are only three places where we need it. What do think (see [this commit](https://github.com/grafana/mimir/pull/8528/commits/c56260e201acc70c0666cb16c68f95777d65706c))?

BTW, it seems that lazy-loading in `LabelValues` is redundant. The function always accesses `b.indexHeader` on the start ([ref](https://github.com/grafana/mimir/blob/ddea70bd87fa40ad8654154280276de631fccea6/pkg/storegateway/bucket.go#L1639)). I've kept it for now.

Reply:
> How about we segregate the creation of the `indexReader` and the warm-up of the blocks index header — there are only three places where we do that. What do think (see [this commit](https://github.com/grafana/mimir/pull/8528/commits/c56260e201acc70c0666cb16c68f95777d65706c))?

yeah that also sounds good. Buut can we still call it `loadIndexReader` instead of warm up? We don't use the warm-up term anywhere else. `loadIndexReader` or `ensureLoadedIndexReader` look ok to me


> BTW, it seems that lazy-loading in `LabelValues` is redundant. The function always calls `b.indexHeader.LabelValuesOffsets()` on the start ([ref](https://github.com/grafana/mimir/blob/ddea70bd87fa40ad8654154280276de631fccea6/pkg/storegateway/bucket.go#L1639)). I've kept it for now.

i realized that's there so that we can record lazy loading duration separately. Otherwise it is conflated with the duration of `LabelValuesOffsets()` and isn't tracked. If we warm up the reader before we invoke `indexHeader.LabelValuesOffsets()`, then this should still work.